client/TracyProfiler.cpp

```diff
@@ -2907,8 +2915,195 @@ void Profiler::HandleSymbolQuery( uint64_t symbol )
+        mapping.end_address = end_addr;
+        mapping.perm_r = perm[0] == 'r';
+        mapping.perm_w = perm[1] == 'w';
+        mapping.perm_x = perm[2] == 'x';
+    }  
+    fclose( file );
+    return result;
+}
+
+// Takes as input an address range [start_address, end_address) and a
+// known vector `mappings`, and returns as output-params a range
+// of iterators into this `mappings` vector, containing the mappings that
+// intersect that address range.
+// Returns true if such a range of iterators exist, false otherwise
+// (that is, if the input address range can't be covered by known mappings).
+bool GetMappingsRange(MappingInfoVector& mappings,
+                      std::uintptr_t start_address,
+                      std::uintptr_t end_address,
+                      MappingInfoIterator* start_iterator,
+                      MappingInfoIterator* end_iterator)
+{
+    // Find the first mapping intersecting the requested address range.
+    // We assume mappings to be sorted by address, as /proc/self/maps seems to be.
+    // Construct a MappingInfo just for the purpose of using std::lower_bound.
+    // (It's an abuse of MappingInfo since this address range is not necessarily
+    // one mapping).
+    MappingInfo address_range;
+    address_range.start_address = start_address;
+    address_range.end_address = end_address;
+    // Comparison function for std::lower_bound. Returns true if all addresses in `m1`
+    // are lower than all addresses in `m2`.
+    auto Compare = []( const MappingInfo& m1, const MappingInfo& m2 ) {
+        // '<=' because the address ranges are half-open intervals, [start, end).
+        return m1.end_address <= m2.start_address;
+    };
+    MappingInfoIterator it = std::lower_bound( mappings.begin(), mappings.end(), address_range, Compare );
+    *start_iterator = it;
+    std::uintptr_t remainder_start_address = start_address;
+    // Loop over all contiguous mappings intersecting the requested address range.
+    while( it != mappings.end() && it->start_address < end_address ) {
+        if (it->start_address > remainder_start_address) {
+            // The current mapping starts after the remainder of the address
+            // range that we needs to cover, so some of that address range
+            // isn't covered by any mapping.
+            return false;
+        }
+        remainder_start_address = it->end_address;
+        ++it;
+    }
+    // Some of the address range isn't covered by any mapping.
+    if (remainder_start_address < end_address) return false;
+    *end_iterator = it;
+    if (*end_iterator < *start_iterator) {
+        abort();
+    }
+    if (*end_iterator > (*start_iterator) + 1) {
+        abort();
+    }
+    return true;
+}
+                      
+
+// Attempts to make the specified `mapping` readable if it isn't already.
+// Returns true if and only if the mapping is readable.
+bool EnsureReadable( MappingInfo* mapping )
+{
+    if( mapping->perm_r )
```

Comment:
The function interface tells me that `mapping` may be `nullptr` here. Change it to a reference to ensure a valid object is always passed to the function.

Reply:
done